From fd9d2e18a01ec97201d5613d21a786b86900ace0 Mon Sep 17 00:00:00 2001
From: Koji Matsuoka <koji.matsuoka.xm@renesas.com>
Date: Mon, 9 Apr 2018 23:13:01 +0700
Subject: [PATCH 337/442] drm: rcar-du: Fix VDSR register calculation

When the setting value of VDSR register is a negative value,
The setting value of VDSR register is rounded at 1 by
specification of H/W.

Signed-off-by: Koji Matsuoka <koji.matsuoka.xm@renesas.com>
Signed-off-by: Hung Dong <hung.dong.xd@rvc.renesas.com>
Signed-off-by: Duy Dang <duy.dang.yb@renesas.com>
---
 drivers/gpu/drm/rcar-du/rcar_du_crtc.c | 14 +++++++++-----
 1 file changed, 9 insertions(+), 5 deletions(-)

diff --git a/drivers/gpu/drm/rcar-du/rcar_du_crtc.c b/drivers/gpu/drm/rcar-du/rcar_du_crtc.c
index a288c51..d6d2076 100644
--- a/drivers/gpu/drm/rcar-du/rcar_du_crtc.c
+++ b/drivers/gpu/drm/rcar-du/rcar_du_crtc.c
@@ -114,6 +114,7 @@ static void rcar_du_crtc_set_display_timing(struct rcar_du_crtc *rcrtc)
 	u32 value;
 	u32 escr;
 	u32 div;
+	int vdsr, vder;
 
 	/* Compute the clock divisor and select the internal or external dot
 	 * clock based on the requested frequency.
@@ -162,11 +163,14 @@ static void rcar_du_crtc_set_display_timing(struct rcar_du_crtc *rcrtc)
 					mode->hsync_start - 1);
 	rcar_du_crtc_write(rcrtc, HCR,  mode->htotal - 1);
 
-	rcar_du_crtc_write(rcrtc, VDSR, mode->crtc_vtotal -
-					mode->crtc_vsync_end - 2);
-	rcar_du_crtc_write(rcrtc, VDER, mode->crtc_vtotal -
-					mode->crtc_vsync_end +
-					mode->crtc_vdisplay - 2);
+	vdsr = mode->crtc_vtotal - mode->crtc_vsync_end - 2;
+	vder = mode->crtc_vtotal - mode->crtc_vsync_end + mode->crtc_vdisplay - 2;
+	if (vdsr < 1) {
+		vder = vder - vdsr + 1;
+		vdsr = 1;
+	}
+	rcar_du_crtc_write(rcrtc, VDSR, vdsr);
+	rcar_du_crtc_write(rcrtc, VDER, vder);
 	rcar_du_crtc_write(rcrtc, VSPR, mode->crtc_vtotal -
 					mode->crtc_vsync_end +
 					mode->crtc_vsync_start - 1);
-- 
2.7.4

